### DISCUSSION TITLE
How to access a variable from inline assembly?

### DISCUSSION THREAD

**Discussion by erlendnagel** [paulscottrobson/neo6502-firmware#383] [Q&A] [answered]
I am trying to change the "room" variable using this piece of code, but somehow it leaves "room" unchanged. Any suggestions as to what I am not getting? This is the room calculator from the original Pitfall by  the way. It is not speed critical, but I thought it fun to copy the original logic.

        room = $C4 : ' starting value
' room' = room << 1 | (bit3 + bit4 + bit5 + bit7)
	for pass = 0 to 1
		o = pass * 1
		p = roomnext
.loop_next
		lda room
		asl
		eor room
		asl
		eor room
		asl
		asl
		eor room
		asl
		rol room
		dex
		bpl loop_next
		rts
	next
	sys roomnext

**Comment by paulscottrobson** [comment, marked as answer]
What it's doing is working as an assembler ; so it's working on memory location $C4

**Comment by Eshu43** [comment]
Inline Assembly Syntax in GCC
GCC inline assembly allows you to write assembly code within your C or C++ programs. The basic syntax is:
Check the code - asm("assembly code" : output operands : input operands : clobbers);

Example: Accessing a Variable
Let’s say you have a variable in your C code and you want to manipulate it using inline assembly. Here’s an example:
Check the code - 
#include <stdio.h>

int main() {
    int var = 10;
    int result;

    asm ("movl %1, %%eax;"
         "addl $5, %%eax;"
         "movl %%eax, %0;"
         : "=r" (result)    /* output operands */
         : "r" (var)       /* input operands */
         : "%eax"          /* clobbered registers */
    );

    printf("Result: %d\n", result);
    return 0;
}

Explanation:
Assembly Code ('asm'):

Check the code -
"movl %1, %%eax;"  // Move the value of input operand (var) into EAX register
"addl $5, %%eax;"  // Add 5 to the value in EAX register
"movl %%eax, %0;"  // Move the value of EAX register to output operand (result)

Output Operands:

"=r" (result): This tells the compiler that the output operand is a general-purpose register ('r'), and its value should be stored in the 'result' variable.

Input Operands:

"r" (var): This tells the compiler that the input operand is a general-purpose register ('r'), and it should take the value from the 'var 'variable.

Clobbered Registers:

"%eax": This informs the compiler that the EAX register is modified by the assembly code, so it won't assume the value in EAX is preserved across this 'asm' block.

Detailed Steps:
Declare Variables: Define your variables in C.
Write Assembly Code: Place the assembly instructions inside the 'asm' block.
Define Operands:
Output Operands: Specify where the results of the assembly instructions should be stored.
Input Operands: Specify the input variables or constants.

Specify Clobbers: List the registers that are modified within the asm block to avoid unexpected behavior.
